src/engine/strat_engine/metadata.rs

```diff
@@ -4,262 +4,416 @@
+
+        self.mdas[older_region].last_updated = Some(*time);
+        self.mdas[older_region].used = data.len() as u64;
+
+        Ok(())
     }
 
-    /// Write SigBlock to buf at offset.
-    pub fn write(&self, buf: &mut [u8], offset: usize) {
-        buf[offset + 4..offset + 20].clone_from_slice(STRAT_MAGIC);
-        LittleEndian::write_u64(&mut buf[offset + 20..offset + 28], *self.total_size);
-        buf[offset + 32..offset + 64]
-            .clone_from_slice(self.pool_uuid.simple().to_string().as_bytes());
-        buf[offset + 64..offset + 96]
-            .clone_from_slice(self.dev_uuid.simple().to_string().as_bytes());
-        self.mda.write(buf, offset + 96);
-        LittleEndian::write_u32(&mut buf[offset + 160..offset + 164],
-                                *self.mda_sectors as u32);
-        LittleEndian::write_u32(&mut buf[offset + 164..offset + 168],
-                                *self.reserved_sectors as u32);
-
-        let hdr_crc = crc32::checksum_ieee(&buf[offset + 4..offset + SECTOR_SIZE as usize]);
-        LittleEndian::write_u32(&mut buf[offset..offset + 4], hdr_crc);
+    pub fn load_state(&self, f: &mut File) -> EngineResult<Option<Vec<u8>>> {
+        let newer_region = self.newer();
+        let mda = &self.mdas[newer_region];
+
+        let mut load_json = |region| {
+            let offset = BDA_STATIC_HDR_SIZE + (region as u64 * *self.region_size * SECTOR_SIZE) +
+                         MDA_REGION_HDR_SIZE as u64;
+            try!(f.seek(SeekFrom::Start(offset)));
+            mda.load_region(f)
+        };
+
+        Ok(load_json(newer_region)
+            .unwrap_or_else(|_| load_json(newer_region + 2).unwrap_or_else(|_| None)))
     }
 
-    /// Read SigBlock from buf at offset.
-    pub fn read(buf: &[u8], offset: usize, total_size: Sectors) -> Result<SigBlock, String> {
-        if &buf[offset + 4..offset + 20] != STRAT_MAGIC {
-            return Err("not a Stratis device".into());
+    pub fn older(&self) -> usize {
+        match self.mdas[0].last_updated.cmp(&self.mdas[1].last_updated) {
+            Ordering::Less => 1,
+            Ordering::Greater => 0,
+            Ordering::Equal => 1,
         }
+    }
 
-        let crc = crc32::checksum_ieee(&buf[offset + 4..offset + SECTOR_SIZE as usize]);
-        if crc != LittleEndian::read_u32(&buf[offset..offset + 4]) {
-            return Err("header CRC invalid".into());
+    pub fn newer(&self) -> usize {
+        match self.older() {
+            0 => 1,
+            1 => 0,
+            _ => panic!("invalid val from older()"),
         }
+    }
+}
 
-        let pool_uuid = try!(Uuid::parse_str(from_utf8(&buf[offset + 32..offset + 64]).unwrap())
-            .map_err(|_| "invalid pool uuid"));
-        let dev_uuid = try!(Uuid::parse_str(from_utf8(&buf[offset + 64..offset + 96]).unwrap())
-            .map_err(|_| "invalid dev uuid"));
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct MDAHeader {
```

Comment:
Have you tried not deriving Eq and PartialEq here. I don't think they're used, and they do not seem like a good idea.